### PR TITLE
Added additional debug info (WebView/Hardware)

### LINKED ISSUE #8234: Add device information (hardware manufacturer and device model number) to Debug Info copied to clipboard in app
It occurs to me that we should probably add Vendor and Model (the device info) to the Debug Info in the app that people copy/paste here, it would be useful

_Originally posted by @mikehardy in https://github.com/ankidroid/Anki-Android/issues/8227#issuecomment-799634011_

Further details:
- "Debug Info" is what we instruct users to copy from the app here https://docs.ankidroid.org/help.html
- similar issue is #8142 (which has a PR I think, and shows where the code is)

### PR DESCRIPTION
**Pull request by 2tanayk**
## Purpose / Description
Added additional and useful debug info -device manufacturer, model, hardware and webview user agent.

## Fixes
Fixes #8234 

## Approach
Used the android os class `Build.` to extract manufacturer, model and hardware information
Used an anonymous instance of the `Webview` class which is `new WebView(this).getSettings().getUserAgentString()` to extract the webview user agent

## How Has This Been Tested?
Tested on an actual android device which runs on Android 10(API 29)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration (SDK version(s), emulator or physical, etc)

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java [hunk 1]
```diff
@@ -221,6 +221,10 @@ public String copyDebugInfo() {
 
         String debugInfo = "AnkiDroid Version = " + VersionUtils.getPkgVersionName() + "\n\n" +
                 "Android Version = " + Build.VERSION.RELEASE + "\n\n" +
+                "Manufacturer = " + Build.MANUFACTURER + "\n\n"+
+                "Model = " + Build.MODEL + "\n\n"+
+                "Hardware = " + Build.HARDWARE+"\n\n"+
+                "Webview User Agent = "+new WebView(this).getSettings().getUserAgentString()+ "\n\n"+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Info.java:227]:
```suggestion
                "WebView User Agent = "+ new WebView(this).getSettings().getUserAgentString()+ "\n\n"+
```

Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Info.java:227]: Could you extract this to a method, and surround it with a `try... catch(Throwable) { AnkiDroidApp.sendExceptionReport()... return null}`. This is somewhere that we don't want a crash for any purpose

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Thanks! 2 minor changes, then good to go

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java (hunk 1 above)
Comment (2tanayk) [reply AnkiDroid/src/main/java/com/ichi2/anki/Info.java:227]: sure

Comment (2tanayk) [reply AnkiDroid/src/main/java/com/ichi2/anki/Info.java:227]: The whole debug info string right??

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/Info.java:227]: Just `new WebView(this).getSettings().getUserAgentString()`, the others can't fail

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java [hunk 2]
```diff
@@ -221,6 +221,10 @@ public String copyDebugInfo() {
 
         String debugInfo = "AnkiDroid Version = " + VersionUtils.getPkgVersionName() + "\n\n" +
                 "Android Version = " + Build.VERSION.RELEASE + "\n\n" +
+                "Manufacturer = " + Build.MANUFACTURER + "\n\n"+
+                "Model = " + Build.MODEL + "\n\n"+
+                "Hardware = " + Build.HARDWARE+"\n\n"+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Info.java:226]:
```suggestion
                "Hardware = " + Build.HARDWARE + "\n\n"+
```

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java (hunk 1 above)
Comment (2tanayk) [reply AnkiDroid/src/main/java/com/ichi2/anki/Info.java:227]: ok got it!

**Review by david-allison-1 — APPROVED** [review]
Looks good to me, thanks!

**Comment by 2tanayk** [comment]
> Looks good to me, thanks!

some JUnit test is failing,why is that?

**Comment by david-allison-1** [comment]
Likely flakiness/nondeterminism in our unit tests - I haven't been through to fix them, and I think the migration from TravisCI to GitHub actions brought out a few issues

Arthur has a PR waiting which should fix the issue once and for all, and I haven't had the time to truly think through the design

**Comment by 2tanayk** [comment]
Any updates regarding this ?Is everything fine?

**Comment by mikehardy** [comment]
I have updated the PR description to include `Fixes #issuenumber` vs just `#issuenumber`, adding the `Fixes` as we request in the PR template is important because that is how GitHub knows to related the PR and issue in a way that closes the issue and cross-links them more strongly.

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java [hunk 3]
```diff
@@ -218,9 +218,20 @@ public String copyDebugInfo() {
             Timber.w(e, "Unable to detect Rust Backend");
         }
 
+        String webviewUserAgent = null;
+        try {
+            webviewUserAgent = getWebviewUserAgent();
+        } catch (Throwable e) {
+            AnkiDroidApp.sendExceptionReport(e, "Info::copyDebugInfo()", "some issue occured while extracting webview user agent");
+            return null;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/Info.java:226]: why would we return in this case, without returning the rest of the information ?
None of the other try/catch blocks return and so in no other case will this method return null, meaning that now consumers have to null check it whereas before it was safe.

This catch block should not return, and definitely should not return null

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java [hunk 4]
```diff
@@ -218,9 +218,20 @@ public String copyDebugInfo() {
...
+        }
 
         String debugInfo = "AnkiDroid Version = " + VersionUtils.getPkgVersionName() + "\n\n" +
                 "Android Version = " + Build.VERSION.RELEASE + "\n\n" +
+                "Manufacturer = " + Build.MANUFACTURER + "\n\n" +
+                "Model = " + Build.MODEL + "\n\n" +
+                "Hardware = " + Build.HARDWARE+"\n\n" +
+                "Webview User Agent = "+ webviewUserAgent +
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/Info.java:234]:
```suggestion
                "Webview User Agent = " + webviewUserAgent +
```

spacing nit: https://github.com/ankidroid/Anki-Android/wiki/Code-style#white-space-should-be-used-in-the-following-cases -> "after and before operators"

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java [hunk 5]
```diff
@@ -236,6 +247,9 @@ public String copyDebugInfo() {
         return debugInfo;
     }
 
+    private String getWebviewUserAgent() {
+        return new WebView(this).getSettings().getUserAgentString()+ "\n\n";
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/Info.java:251]:
```suggestion
        return new WebView(this).getSettings().getUserAgentString() + "\n\n";
```
https://github.com/ankidroid/Anki-Android/wiki/Code-style#white-space-should-be-used-in-the-following-cases

**Review by mikehardy — CHANGES_REQUESTED** [review]
one control flow issue, two spacing nits, excited to get this expanded info from users though!

**Comment by 2tanayk** [comment]
Will get it done as soon as possible sir✌️

**Comment by 2tanayk** [comment]
done @mikehardy sir!

**Review by david-allison-1 — COMMENTED** [review]
As discussed on Discord - needs rebasing onto master and removing the merge commit

**Comment by mikehardy** [comment]
Yeah, sorry - the commit history is non-reviewable at the moment. Once cleaned up I can re-check

**Comment by 2tanayk** [comment]
> Yeah, sorry - the commit history is non-reviewable at the moment. Once cleaned up I can re-check

you mean I need to squash all the commits into one ?

**Comment by mikehardy** [comment]
When I open the "Commits" tab on the PR I should see nothing but clean commits that implement your changes only, no unrelated commits

When I open the "Files Changed" tab on the PR I should see nothing but an easily readable diff containing nothing but the actual changes you needed to make to implement the improvement

There are many ways to do this with git, I don't have the time to lead you through them, but choose one that meets my expectations of a clean diff as described here, and I can review it

**Comment by 2tanayk** [comment]
@mikehardy  sir how about a fresh/new PR?

**Comment by mikehardy** [comment]
That is sometimes the best option, although I think there is value for you - as a person learning git - in figuring out how to actually fix this using `git rebase`. It won't be ...fun, but you will have much stronger control over git afterwards. Your choice

**Comment by 2tanayk** [comment]
> That is sometimes the best option, although I think there is value for you - as a person learning git - in figuring out how to actually fix this using `git rebase`. It won't be ...fun, but you will have much stronger control over git afterwards. Your choice

Yes I get that!

**Comment by 2tanayk** [comment]
> Yeah, sorry - the commit history is non-reviewable at the moment. Once cleaned up I can re-check

> When I open the "Commits" tab on the PR I should see nothing but clean commits that implement your changes only, no unrelated commits
> 
> When I open the "Files Changed" tab on the PR I should see nothing but an easily readable diff containing nothing but the actual changes you needed to make to implement the improvement
> 
> There are many ways to do this with git, I don't have the time to lead you through them, but choose one that meets my expectations of a clean diff as described here, and I can review it

sir if you want the file changed/diff then you can view it on github using a two-dot diff comparison like this:

https://github.com/2tanayk/Anki-Android/compare/e8141a2..38d670a

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java [hunk 6]
```diff
@@ -218,9 +218,19 @@ public String copyDebugInfo() {
...
+            webviewUserAgent = getWebviewUserAgent();
+        } catch (Throwable e) {
+            AnkiDroidApp.sendExceptionReport(e, "Info::copyDebugInfo()", "some issue occured while extracting webview user agent");
+        }
 
         String debugInfo = "AnkiDroid Version = " + VersionUtils.getPkgVersionName() + "\n\n" +
                 "Android Version = " + Build.VERSION.RELEASE + "\n\n" +
+                "Manufacturer = " + Build.MANUFACTURER + "\n\n" +
+                "Model = " + Build.MODEL + "\n\n" +
+                "Hardware = " + Build.HARDWARE+"\n\n" +
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Info.java:232]:
```suggestion
                "Hardware = " + Build.HARDWARE+ "\n\n" +
```

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java [hunk 7]
```diff
@@ -236,6 +246,9 @@ public String copyDebugInfo() {
         return debugInfo;
     }
 
+    private String getWebviewUserAgent() {
+        return new WebView(this).getSettings().getUserAgentString() + "\n\n";
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Info.java:250]: Nitpick/Implementer's choice: code style

I feel that the try/catch would be better inside this method, this allows us to reduce code in the `copyDebugInfo` method, and move the method call inline to where it's used

**Comment by 2tanayk** [comment]
Now let me know if I've got it right this time :)

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Thanks! Tiny nitpick.

Almost there.

* Could you squash the 3 commits into one

I feel the commit message would be better as:

```
Added additional debug info (WebView/Hardware)

Fixes #8234
```

This provides more context for the commit, the issue number isn't too relevant when looking for context in a long list of issues.

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java (hunk 7 above)
Comment (2tanayk) [reply AnkiDroid/src/main/java/com/ichi2/anki/Info.java:250]: Yeah thats what I really wanted to do but it gives a compile time error (Java doesn't risk return in non-void functions) and hence I will have to return something else if the try block throws an exception(probably null) which makes the code unsafe again.

Comment (2tanayk) [reply AnkiDroid/src/main/java/com/ichi2/anki/Info.java:250]: I will squash the commits for now

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/Info.java:250]: You can return null, that's what we're doing implicitly here anyway

Comment (2tanayk) [reply AnkiDroid/src/main/java/com/ichi2/anki/Info.java]: ok did that

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java [hunk 8]
```diff
@@ -236,6 +240,14 @@ public String copyDebugInfo() {
         return debugInfo;
     }
 
+    private String getWebviewUserAgent() {
+        try {
+            return new WebView(this).getSettings().getUserAgentString() + "\n\n";
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Info.java:245]: Move the `+ "\n\n";` to the `"Webview User Agent =` line, so it's applied to both null and the valid value

File: AnkiDroid/src/main/java/com/ichi2/anki/Info.java [hunk 9]
```diff
@@ -218,9 +218,13 @@ public String copyDebugInfo() {
             Timber.w(e, "Unable to detect Rust Backend");
         }
 
-
+        String webviewUserAgent = getWebviewUserAgent();
         String debugInfo = "AnkiDroid Version = " + VersionUtils.getPkgVersionName() + "\n\n" +
                 "Android Version = " + Build.VERSION.RELEASE + "\n\n" +
+                "Manufacturer = " + Build.MANUFACTURER + "\n\n" +
+                "Model = " + Build.MODEL + "\n\n" +
+                "Hardware = " + Build.HARDWARE+ "\n\n" +
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Info.java:226]: And we're done after this, thanks for the patience

```suggestion
                "Hardware = " + Build.HARDWARE + "\n\n" +
```

**Review by david-allison-1 — COMMENTED** [review]
Last one!

One last thing!

Note that the PR title isn't the commit message title

You can change the commit message title via a `reword` operation in a rebase.

Alternately, Android Studio lets you do this with a UI:

image

**Comment by 2tanayk** [comment]
I think all is taken care of ,check

**Review by david-allison-1 — APPROVED** [review]
Looks good to me, thanks!

**Review by mikehardy — APPROVED** [review]
good to go! Thanks

**Comment by mikehardy** [comment]
Hi there!

Just a friendly notice (one per person for PRs merged in March, regardless of PR quantity) that we try to process OpenCollective payments monthly - it's time for March 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

Thanks!

**Comment by mikehardy** [comment]
@2tanayk I was able to just immediately diagnose a user problem as an out-of-date webview based on this extra debugging information, and it fixed the user immediately when they updated. I love seeing things like this pay off! Thanks again

**Comment by 2tanayk** [comment]
Thanks @mikehardy means a lot to me and my semester almost over,look to start contributing again!

**Pull request merged**